### PR TITLE
Begin api transition from changeset Ids to changeset index.

### PR DESCRIPTION
**Pull request by kabentley**
`ChangesetId`s  are essentially a random string of numbers. `ChangesetIndex` is a small integer representing the relative position on the timeline. It is easy to compare two `ChangeseIndex`es or sort an array of them, but both are stable identifiers of a changeset. Our api generally uses ChangesetIds, but that makes everything unnecessarily hard. Instead, use index for all access to changesets from iModelHub. 

Note that the iModel currently (only) holds `parentChangesetId`, and we use that value in `IModelRpcProps`. It would be nice to convert that to index too.

Likewise, checkpoints are currently named by changesetId. They should also use changesetIndex.

Also note that IModelHub currently does not guarantee there won't be holes (missing values) in the index sequence. That's unfortunate and should be fixed, but we don't rely that index values always increase by 1.

### CODE REVIEW

**Review by pmconne — COMMENTED** [review]
Looks reasonable to me but I think @ramanujam-raman is better qualified to review.

**Comment by kabentley** [comment]
> Looks good to me. I didn't see any updates to ChangedElements.test.ts (part of backend integration tests). As long as that's passing, my end of things will be ok.

That's because those tests download *all* changesets, so they didn't need to change. But, i see that you use `pushDate` to sort the order of changesets. You should switch to using `index` and you should switch your entire api to be index-based.

**Comment by diegopinate** [comment]
> > Looks good to me. I didn't see any updates to ChangedElements.test.ts (part of backend integration tests). As long as that's passing, my end of things will be ok.
> 
> That's because those tests download _all_ changesets, so they didn't need to change. But, i see that you use `pushDate` to sort the order of changesets. You should switch to using `index` and you should switch your entire api to be index-based.

Yes, this will definitely help in multiple areas where I needed to query the changeset Ids and sort them. I will revisit that after your PR is merged, added action item. Thanks.

File: core/backend/src/BackendHubAccess.ts [hunk 1]
```diff
@@ -187,7 +193,7 @@ export interface BackendHubAccess {
   queryAllLocks: (arg: BriefcaseDbArg) => Promise<LockProps[]>;
 
   /** release all currently held locks */
-  releaseAllLocks: (arg: BriefcaseIdArg) => Promise<void>;
+  releaseAllLocks: (arg: BriefcaseIdArg & ChangesetIndexArg) => Promise<void>;
```
[the remark this answers is not in the document]
Comment (kabentley) [reply core/backend/src/BackendHubAccess.ts:215]: you're right, i just added it. 

But I need another PR for the reworked lock stuff.

**Review by wgoehrig — APPROVED** [review]

File: core/backend/src/ConcurrencyControl.ts [hunk 2]
```diff
@@ -551,8 +553,8 @@ export class ConcurrencyControl {
   }
 
   /** Abandon any *pending* requests for locks or codes.
-   * This is called automatically by BriefcaseDb.abandonChanges.
-  */
+ * This is called automatically by BriefcaseDb.abandonChanges.
```
Comment (calebmshafer) [inline core/backend/src/ConcurrencyControl.ts:556]: nit: formatting seems to have gotten messed up here.

File: core/backend/src/BackendHubAccess.ts [hunk 3]
```diff
@@ -59,28 +72,37 @@ export interface ChangesetFileProps extends ChangesetProps {
...
+  /** index of the first changeset */
+  first: ChangesetIndex;
+  /** index of last changeset. If undefined, all changesets after first are returned. */
+  end?: ChangesetIndex;
+}
+
+/** The scope of a lock.
+ * @public
+ */
+export enum LockScope {
```
Comment (calebmshafer) [inline core/backend/src/BackendHubAccess.ts:88]: Do you want to add something to the NextVersion.md about the change to LockScope and LockLevel api changes or do you have other changes coming?

Comment (kabentley) [reply core/backend/src/BackendHubAccess.ts:88]: this is all internal for now (even though that enum is marked @public). I want to work on my next PR which will be to rework concurrencyControl and locking in general.

Comment (calebmshafer) [reply core/backend/src/BackendHubAccess.ts:88]: sounds good.

**Review by calebmshafer — APPROVED** [review]

**Pull request merged**